### ISSUE TITLE
Temp

### ISSUE THREAD

**Issue by hy21116** [hy21116/EfficientDet-pytorch#9]

**Comment by hy21116** [comment]
- 일단 1차 구현을 위해서 count 기준은 업로드 한 feed 개수로 통일
- [ ] photo? image? 하나로 통일
- [ ] search 화면 만들기

### User
![image](https://user-images.githubusercontent.com/13702048/145231617-da3c6c0d-b894-4179-843a-5ae85ce1cf57.png)

- 오늘/금주의
  - feed 개수 어떻게 count? sort?
- 인기 급상승 (** 개발 뒤로 미루기ㅋㅋ) 
  - 좋아요 많은 feed 순서대로

### Problem
![image](https://user-images.githubusercontent.com/13702048/145230814-b6e91b15-9427-48b9-b806-7eb84f5e2adc.png)

- 금주의 HOT 볼더링
  - feedIds 개수?
  - 이번주... 어떻게 count?
- 이 문제는 어떤 맛일까? (비슷한 레벨의 유저들이 많이 푼 문제 추천)
- 내가 시도했지만 못 푼 문제 (** 개발 뒤로 미루기ㅋㅋ)

**Comment by hy21116** [comment]
## Overfitting

![image](https://user-images.githubusercontent.com/13702048/188451863-b631492f-aa2e-43f6-b388-3f8a01d1829a.png)

- Train data를 과하게 학습. Test data(실제 데이터)에 대한 오차 증가 
    - overfitting이려면 train_loss는 줄고 accuracy는 떨어져야 함
- 1) overfitting은 데이터가 적을수록 많이 발생. 종종 복잡한 경우에도 발생하나 데이터가 많을 경우 복잡하더라도 발생하지 않음
- 2) 매개변수가 많고 표현력이 높은 모델인 경우
    - parameter 多
    - parameter 값의 범위 넓은 경우

### 해결방안
#### 1. Data augmentation
https://hleecaster.com/ml-overfitting/
- 편향되지 않은 data
  - variance와 bias: https://acdongpgm.tistory.com/53
- 클래스당 이미지 갯수 중요
   https://www.inflearn.com/questions/214098
   ```
   추가적으로, 전체 클래스가 몇개 인지요? 
   그러니까 클래스가 10개 인데, validation이 200장인지, 
   아님 클래스가 100개 인데 validation이 200장인지 궁금하군요. 
   후자의 경우에는 잘못되면 metric/loss가 널뛰기를 할 수 있습니다.
   ```

#### 2. Regularization
https://umbum.dev/222
https://jwlee010523.tistory.com/entry/%EA%B3%BC%EC%A0%81%ED%95%A9Overfitting
https://box-world.tistory.com/16
- weight decay (L2 regularization)
- dropout

#### 3. Transfer learning

- 
-

**Comment by hy21116** [comment]
train loss도 증가하고 accuracy도 증가..?

#### train loss
```
(img1, img2)                ------- [128, 3, 512, 512]
out1 = self.forward_one(img1)  ------- [128, 4096] 
out2 = self.forward_one(img2)
dis = torch.abs(out1 - out2)
self.out = nn.Linear(4096, 1)
output = self.out(dis)          ------- [128, 1]

loss_fn = torch.nn.BCEWithLogitsLoss(size_average=True)
loss = loss_fn(output, label)
loss_val += loss.item()
loss.backward()
```
- `loss_val`은 `show_every` iter마다 reset
- `BCEWithLogitsLoss`는

#### accuracy
```
(img1, img2)                              ------- [20, 3, 512, 512]
# same with above
output = self.out(dis)                      ------- [20, 1]
pred = np.argmax(output.data.cpu().numpy())
if pred == 0:
    right += 1
else: error += 1
accuracy = right*1.0/(right+error)
```
- 20-way: 하나 선택한 뒤, 랜덤하게 구성된 20가지의 선택지

![image](https://user-images.githubusercontent.com/13702048/188522698-5e945dba-178e-4946-b9a1-57c4bc336fdd.png)


----

- train/test_loader 각각 몇 개씩인지?
  - omniglot: 50개의 알파벳, 각 15개~40개의 글자
  - 30개의 background set, 20개의 evaluation set

**Comment by hy21116** [comment]
#### 1. dataset 문제
- 각 문제 당 사진 수를 높여보자 (= omniglot)
- validation set을 40% 수준으로 높여보자 (= omniglot)
- 

#### 2. 모델에 비해 문제가 너무 쉬운 것 아닌지?
- false negative를 통해 확인해보자

**Comment by hy21116** [comment]
## TODO 

### 1. Dataset 문제

- omniglot
  - 50개의 알파벳, 각 15~40개의 글자
  - 30개의 background set, 20개의 evaluation set
- https://www.inflearn.com/questions/214098
  - 검증을 학습데이터의 40% 정도로 할당해야 한다
  - 각 클래스 당 이미지 갯수 영향 있음
       ```
       추가적으로, 전체 클래스가 몇개 인지요? 
       그러니까 클래스가 10개 인데, validation이 200장인지, 
       아님 클래스가 100개 인데 validation이 200장인지 궁금하군요. 
       후자의 경우에는 잘못되면 metric/loss가 널뛰기를 할 수 있습니다.
       ```

- [ ] 각 문제당 사진 수 증가 
- [ ] evaluation set을 40% 수준으로 높이기 (30:20 / 40:25)


### 2. 모델에 비해 문제가 너무 쉬운 것 아닌지?
- [ ] train_loss/accuracy 급증하는 구간 디버깅
- [ ] false negative 사례들 visualization
- [ ] 더 작은 모델로 돌려보기

### 3. Hyper-parameter 
- [ ] Learning rate scheduler: SGD, Adam

### 4. Data augmentation
